Fix the build for NLL

SimonSapin opened the pull request:
Test with `RUSTFLAGS="-Zborrowck=mir -Ztwo-phase-borrows" cargo build`

https://internals.rust-lang.org/t/help-us-get-non-lexical-lifetimes-nll-over-the-finish-line/7807/7

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/21356)
<!-- Reviewable:end -->

SimonSapin committed:
Fix the build for NLL

Test with `RUSTFLAGS="-Zborrowck=mir -Ztwo-phase-borrows" cargo build`

https://internals.rust-lang.org/t/help-us-get-non-lexical-lifetimes-nll-over-the-finish-line/7807/7

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/webglrenderingcontext.rs, components/script/dom/imagedata.rs, components/script/dom/window.rs
 * @kichjang: components/script/dom/webglrenderingcontext.rs, components/script/dom/imagedata.rs, components/script/dom/window.rs
 * @emilio: components/layout/table.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify layout and script code, but no tests are modified. Please consider adding a test!

emilio reviewed (approved):
Looks good :)

KiChjang commented on components/layout/table.rs:
```diff
@@ -448,9 +448,9 @@ impl Flow for TableFlow {
+                        let inline_size = self.column_computed_inline_sizes[index].size.0;
                         self.column_computed_inline_sizes[index].size +=
-                            remaining_inline_size.scale_by(
-                                self.column_computed_inline_sizes[index].size.0 as f32 / total_minimum_size);
```
This is possibly rust-lang/rust#47349, which I'm working on.

SimonSapin replied:
Per https://github.com/rust-lang/rust/issues/47349#issuecomment-362295502 the new borrowck behavior is considered correct. That issue is only still open in order to improve the error message.

SimonSapin commented on components/layout/table.rs:
```diff
@@ -531,7 +531,7 @@ impl Flow for TableFlow {
 
         let iter = TableCellStyleIterator::new(&self);
-        for mut style in iter {
+        for style in iter {
```
Yes, I suppose that they were false negatives for the warning under AST-borrowck.

SimonSapin commented:
Oops I accidentally pushed this to master :/ Thanks @jdm for taking care of reverting. Don’t we have github settings to prevent this from happening, though?

SimonSapin commented:
Per https://github.com/servo/servo/issues/21358 it looks like unrelated changes in `url` fixed some tests, so this PR will need to update expectations.

@bors-servo try

bors-servo commented:
:hourglass: Trying commit e510dea7524f2feb242709cf7f5c4ab1f326c26b with merge 347d90c5c0264d031be912216abcfb7e6eb9799d...

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit ff77763 has been approved by `jdm`

<!-- @bors-servo r=jdm ff777632b4ba31ceb32fa93bbde9e6cbfbef1432 -->

jdm commented:
@bors-servo retry

jdm commented:
@bors-servo force r+

bors-servo commented:
:bulb: This pull request was already approved, no need to approve it again.

bors-servo commented:
:pushpin: Commit ff77763 has been approved by `jdm`

<!-- @bors-servo r=jdm ff777632b4ba31ceb32fa93bbde9e6cbfbef1432 -->

jdm commented:
@bors-servo clean r-

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit ff77763 has been approved by `jdm`

<!-- @bors-servo r=jdm ff777632b4ba31ceb32fa93bbde9e6cbfbef1432 -->

jdm commented:
@bors-servo retry

jdm commented:
@bors-servo r-

jdm commented:
@bors-servo

bors-servo commented:
:pushpin: Commit ff77763 has been approved by `jdm`

<!-- @bors-servo r=jdm ff777632b4ba31ceb32fa93bbde9e6cbfbef1432 -->

jdm commented:
@bors-servo r-

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit ff77763 has been approved by `jdm`

<!-- @bors-servo r=jdm ff777632b4ba31ceb32fa93bbde9e6cbfbef1432 -->

bors-servo commented:
:hourglass: Testing commit ff777632b4ba31ceb32fa93bbde9e6cbfbef1432 with merge 3a42a16edb421630e78741d45b22d17bc423c653...

bors-servo commented:
:broken_heart: Test failed - [android-x86](http://build.servo.org/builders/android-x86/builds/272)

jdm commented:
@bors-servo retry

bors-servo commented:
:hourglass: Testing commit ff777632b4ba31ceb32fa93bbde9e6cbfbef1432 with merge 69a62270796361fe446a75726d6f7c4369ebcb3c...

bors-servo commented:
:broken_heart: Test failed - [android-x86](http://build.servo.org/builders/android-x86/builds/276)

jdm commented:
@bors-servo retry

bors-servo commented:
:hourglass: Testing commit ff777632b4ba31ceb32fa93bbde9e6cbfbef1432 with merge 698102f834c5cef2cf8ae87810c765750b468552...

bors-servo commented:
:broken_heart: Test failed - [android-x86](http://build.servo.org/builders/android-x86/builds/279)

jdm commented:
@bors-servo try=android

Manishearth commented:
@bors-servo ping

bors-servo commented:
:sleepy: I'm awake I'm awake

Manishearth commented:
@bors-servo try=fake

bors-servo commented:
:slightly_frowning_face: There is no try chooser fake for this repo, try one of: wpt, android, build

Manishearth commented:
@bors-servo try=android

Manishearth commented:
@bors-servo try- clean

Manishearth commented:
@bors-servo try=android

Manishearth commented:
@bors-servo r- try-

Manishearth commented:
@bors-servo try=android retry

bors-servo commented:
:hourglass: Trying commit ff777632b4ba31ceb32fa93bbde9e6cbfbef1432 with merge 0faca3dd2c14c6f4d0a7332b19419c5069d3cd93...

bors-servo commented:
:broken_heart: Test failed - [android-x86](http://build.servo.org/builders/android-x86/builds/284)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #20678) made this pull request unmergeable. Please resolve the merge conflicts.

SimonSapin commented:
@bors-servo r=jdm

bors-servo commented:
:pushpin: Commit 5c8352b has been approved by `jdm`

<!-- @bors-servo r=jdm 5c8352b2a488be029accccf94ff1190b3e8d5aa4 -->

bors-servo commented:
:hourglass: Testing commit 5c8352b2a488be029accccf94ff1190b3e8d5aa4 with merge 4a8caaf34537c597ca11496b84ae4c8f8e5ae8ec...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/12516)

jdm commented:
@bors-servo retry
* cleaned out the builder

bors-servo commented:
:hourglass: Testing commit 5c8352b2a488be029accccf94ff1190b3e8d5aa4 with merge 6ee8c28649a91b7d098f7a232a27260aff5cae68...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/5062)

jdm commented:
@bors-servo retry
* #20734

jdm commented:
This continues to be broken on every single android-x86 run.

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/9303)

bors-servo commented:
:broken_heart: Test failed - [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/9441)

jdm commented:
@bors-servo retry
* #20734 #21409

jdm commented:
@bors-servo r-
I keep forgetting about the android problem.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #21386) made this pull request unmergeable. Please resolve the merge conflicts.

SimonSapin commented:
@jdm what’s the android problem?

Maybe I’d find out by trying:

@bors-servo r=jdm

bors-servo commented:
:pushpin: Commit 3b3f6bb has been approved by `jdm`

<!-- @bors-servo r=jdm 3b3f6bb7c9b46cd9f50651ea9a8e833644809af5 -->

bors-servo commented:
:hourglass: Testing commit 3b3f6bb7c9b46cd9f50651ea9a8e833644809af5 with merge 6d5ee781e8fa91be75e3c5c4fa849b87862d7874...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/12547)

jdm commented:
@SimonSapin http://servo-master.servo.org:54856/results/servo/21356 reinforces https://github.com/servo/servo/pull/21356#issuecomment-411902287.

SimonSapin commented:
I’ve backported the NLL fix in the `url` crate as 1.6.1, to avoid bringing in (in this PR) unrelated changes.

There is now the `gstreamer` crate whose current versions fail to build with NLL. This is fixed in git upstream, but not yet on crates.io versions. I’ve asked for a backport: https://github.com/sdroege/gstreamer-rs/commit/dba110e8ad255f7327d9f70cd7565917a303df12

Still, the rest of this PR can be landed without waiting for that.

@bors-servo r=jdm

(Carrying previous r+ since the PR is now a subset of the changes since the previous review.)

bors-servo commented:
:pushpin: Commit 6539b18 has been approved by `jdm`

<!-- @bors-servo r=jdm 6539b1875a47ad50ebd8b614c6b681d2bcb81f15 -->

bors-servo commented:
:hourglass: Testing commit 6539b1875a47ad50ebd8b614c6b681d2bcb81f15 with merge ac832f8ea5116941ff0c27d522edfe8c1dfa2287...

SimonSapin commented:
Same as before:

```
 5:03.86 ERROR Traceback (most recent call last):
  File "/home/servo/buildbot/slave/android-x86/build/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/executors/protocol.py", line 47, in setup
    self.connect()
  File "/home/servo/buildbot/slave/android-x86/build/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/executors/executorservodriver.py", line 97, in connect
    wait_for_service((self.host, self.port), timeout=self.init_timeout)
  File "/home/servo/buildbot/slave/android-x86/build/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/webdriver_server.py", line 265, in wait_for_service
    raise socket.error("Service is unavailable: %s:%i" % addr)
error: Service is unavailable: 127.0.0.1:4444
```

So my assumption that this was due to changes in the URL crate (perhaps in the webdriver server implementation) was not correct. This issue will need more investigation. (getopts might be another lead to follow.)

SimonSapin committed:
Update getopts

*Sigh* https://github.com/rust-lang-nursery/getopts/pull/53#issuecomment-415770100

servo-wpt-sync commented:
Opened new PR for upstreamable changes.

Completed upstream sync of web-platform-test changes at https://github.com/jdm/web-platform-tests/pull/78.

SimonSapin commented:
*Sigh*. `getopts` silently removed support for `--webdriver 4444` (instead of `--webdriver=4444`) in a micro version number increment.

Added `=` in a few places, r?

jdm commented:
@bors-servo r+

bors-servo commented:
:pushpin: Commit 5d40886 has been approved by `jdm`

<!-- @bors-servo r=jdm 5d408862a730dce64b9fcfd3eb4f092f188b5255 -->

bors-servo commented:
:hourglass: Testing commit 5d408862a730dce64b9fcfd3eb4f092f188b5255 with merge dcc8a6e393423fa85fa3d8c8ecbf311dd65b30f0...

SimonSapin commented:
WPT manifest updated.

@bors-servo r=jdm

bors-servo commented:
:pushpin: Commit 1333b6e has been approved by `jdm`

<!-- @bors-servo r=jdm 1333b6ee4513926fa192d438d16771922ce047a9 -->

servo-wpt-sync commented:
Transplanted upstreamable changes to existing PR.

Completed upstream sync of web-platform-test changes at jdm/web-platform-tests#78.

bors-servo commented:
:hourglass: Testing commit 1333b6ee4513926fa192d438d16771922ce047a9 with merge 39551199c7aef349b25f96178f4bc65d0fea554c...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/12331), [android-x86](http://build.servo.org/builders/android-x86/builds/442), [arm32](http://build.servo.org/builders/arm32/builds/12525), [arm64](http://build.servo.org/builders/arm64/builds/12567), [linux-dev](http://build.servo.org/builders/linux-dev/builds/12627), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/9559), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/9558), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/12655), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/5098), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/5098), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/9365), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/9409), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/5170), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/5114), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/8031)
Approved by: jdm
Pushing 39551199c7aef349b25f96178f4bc65d0fea554c to master...

servo-wpt-sync commented:
Error syncing changes upstream. Logs saved in error-snapshot-1535130643349.
